controllers/imagehub/api/v1/repository_webhook.go

```diff
@@ -17,70 +17,131 @@ limitations under the License.
+	if !ok {
+		return errors.New("obj convert Repository is error")
 	}
-	return nil
+	repositorylog.Info("validating create", "name", r.Name)
+	repositorylog.Info("enter checkOption func", "name", r.Name)
+	return v.checkOption(ctx, r)
 }
 
-// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
-func (r *Repository) ValidateUpdate(old runtime.Object) error {
-	repositorylog.Info("validate update", "name", r.Name)
-	if !r.checkSpecName() {
-		return fmt.Errorf("repo name illegal")
+func (v *RepoValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	nr, ok := newObj.(*Repository)
+	if !ok {
+		return errors.New("obj convert Repository is error")
 	}
-	if !r.checkLabels() {
-		return fmt.Errorf("repo lable illegal")
+	or, ok := oldObj.(*Repository)
+	if !ok {
+		return errors.New("obj convert Repository is error")
 	}
-	return nil
+	imagelog.Info("validating update", "name", or.Name)
+	if nr.Spec.Name != or.Spec.Name {
+		return fmt.Errorf("can not change spec.name: %s", string(nr.Spec.Name))
+	}
+	imagelog.Info("enter checkOption func", "name", nr.Name)
+	return v.checkOption(ctx, nr)
 }
 
-// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
-func (r *Repository) ValidateDelete() error {
-	repositorylog.Info("validate delete", "name", r.Name)
+func (v *RepoValidator) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	r, ok := obj.(*Repository)
+	if !ok {
+		return errors.New("obj convert Repository is error")
+	}
+	repositorylog.Info("validating delete", "name", r.Name)
+	repositorylog.Info("enter checkOption func", "name", r.Name)
+	return v.checkOption(ctx, r)
+}
 
-	// TODO(user): fill in your validation logic upon object deletion.
-	return nil
+func (v *RepoValidator) checkOption(ctx context.Context, r *Repository) error {
+	repositorylog.Info("checking label and spec name", "repository name", r.Spec.Name)
+	if !r.checkLabels() || !r.checkSpecName() {
+		return fmt.Errorf("missing labels or repository.Spec.Name is IsLegal: %s", string(r.Spec.Name))
+	}
+	repositorylog.Info("getting org", "org", r.Spec.Name.GetOrg())
+	org := &Organization{}
+	if err := v.Get(ctx, client.ObjectKey{Name: r.Spec.Name.GetOrg()}, org); err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			return fmt.Errorf("organization not exited %s", r.Spec.Name.GetOrg())
+		}
+		return fmt.Errorf("get Organization error %s", r.Spec.Name.GetOrg())
+	}
+	repositorylog.Info("org info", "org", org)
+	repositorylog.Info("getting req from ctx")
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		repositorylog.Info("get request from context error when validate", "repository name", r.Name)
+		return err
+	}
+	repositorylog.Info("checking user", "user", req.UserInfo.Username)
+	// get sa namespace prefix, prefix format is like: "system:serviceaccount:user-system:"
+	namespacePrefix := fmt.Sprintf("%s:%s:", saPrefix, getUserNamespace())
```

Comment:
This `saPrefix` value is shared over multiple webhooks, it should move to standalone const file, like move into `common_types.go`

Reply:
ok